terra_notebook_utils/table.py

```diff
@@ -1,10 +1,131 @@
+        if isinstance(item, dict):
+            row = Row(f"{uuid4()}", item)
+        else:
+            row = Row(*item)
+        column_headers = tuple(sorted(row.attributes.keys()))
+        if column_headers not in self._tsvs:
+            self._tsvs[column_headers] = "\t".join([f"{self.name}_id", *column_headers])
+            self._row_update_request_data[column_headers] = list()
+        self._tsvs[column_headers] += (
+            os.linesep
+            + row.name + "\t"
+            + "\t".join(row.attributes[c] if isinstance(row.attributes[c], str)
+                        else "x"
+                        for c in column_headers)
+        )
+        update_request_data = self._get_row_update_request_data(row)
+        if update_request_data:
+            self._row_update_request_data[column_headers].append((row, update_request_data))
+        self._upload(1024 * 20)
+        return row.name
+
+    def _upload(self, threshold: Optional[int]=None):
+        """
+        Schedule uploads for all TSVs of size equal to or greater than `threshold`.
+        If `threshold` is None, schedule uploads for all TSVs.
+        """
+        for column_headers, tsv in self._tsvs.copy().items():
+            if threshold is None or len(tsv) >= threshold:
+                row_updates = self._row_update_request_data.get(column_headers, list())
+                self.submit(self._do_fiss_upload, tsv, row_updates)
+                del self._tsvs[column_headers]
+                if row_updates:
+                    del self._row_update_request_data[column_headers]
+
+    def _do_fiss_upload(self, tsv: str, row_update_request_data: List[Tuple[Row, List[Dict[str, Any]]]]):
+        fiss.fapi.upload_entities(self._workspace_google_project,
+                                  self._workspace_name,
+                                  tsv,
+                                  model="flexible").raise_for_status()
+        for row, request_data in row_update_request_data:
+            if request_data:
+                self.submit(self._do_fiss_updates, row, request_data)
+
+    def _do_fiss_updates(self, row: Row, request_data: UPDATE_OPS, retry: int=0):
+        try:
+            fiss.fapi.update_entity(self._workspace_google_project,
+                                    self._workspace_name,
+                                    self.name,
+                                    row.name,
+                                    request_data).raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            if 500 == e.response.status_code:
+                # Firecloud occasionally throws 500 errors for succesful update operations.
+                # Check if we get the row we expect, retry otherwise.
+                if 5 > retry:
+                    cur_row = get_row(self.name, row.name)
+                    if row != cur_row:
+                        self.submit(self._do_fiss_updates, row, request_data, retry=retry + 1)
```

Comment:
Should this be incrementing a retry every time instead of subtracting?  I'm not very clear on how this is retrying.

Possibly consider a way to make this work with a `retry` decorator.

Reply:
Not an odd choice to me :)

If this pattern is used more than once we can consider an abstraction.